### PR TITLE
[as4224/as5114] Upgrade kernel to 5.10

### PR DESCRIPTION
**Pull request by brandonchuang**
Signed-off-by: Brandon Chuang <brandon_chuang@edge-core.com>

### CODE REVIEW

File: packages/platforms/accton/arm64/as4224/src/modules-5_10/arm64-accton-as4224-psu.c [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+/* Each client has this additional data
+ */
+struct as4224_psu_data {
+    struct platform_device *pdev;
+    struct device      *hwmon_dev;
+    struct mutex       update_lock;
+    char               valid;        /* !=0 if registers are valid */
+    unsigned long       last_updated; /* In jiffies */
+    u8  status;          /* power_good status read from CPLD */
+    u8  psu_count;
```
Comment (paulmenzel) [inline packages/platforms/accton/arm64/as4224/src/modules-5_10/arm64-accton-as4224-psu.c:58]: Would be nice if the Linux kernel coding style could be followed.

File: packages/platforms/accton/arm64/as4224/src/modules-5_6/arm64-accton-as4224-cpld.c [hunk 2]
```diff
@@ -92,7 +92,7 @@ enum cpld_type {
 };
 
 enum as4224_platform_id {
-    AS4224_48X,
+    AS5114_48X,
```
Comment (paulmenzel) [inline packages/platforms/accton/arm64/as4224/src/modules-5_6/arm64-accton-as4224-cpld.c:95]: Could this be a separate commit with an explanation?

File: packages/base/any/kernels/5.10-lts/patches/0012-accton-as4224.patch [hunk 3]
```diff
@@ -81,33 +81,37 @@ index 000000000..d38b19668
 +			It is currently not required for Amazon 'ethtool -m' support but it is intended for future use.
 +			Can be skipped in this stage.
 +		*/
-+		los-gpio = <&gpio_i2c 0 GPIO_ACTIVE_HIGH>;
++		los-gpio = <&gpio_i2c 0 GPIO_ACTIVE_LOW>;
```
Comment (paulmenzel) [inline packages/base/any/kernels/5.10-lts/patches/0012-accton-as4224.patch:84]: Please make this change in a separate commit with the appropriate description.

**Review by paulmenzel — COMMENTED** [review]
Thank you for submitting the changes. Currently, it’s hard to review though, as it’s only one big commit, and it looks like there are small changes in the patches, which I do not fully understand, how they are related to updating the Linux kernel version.

Also, a description would be nice, what new features the Linux kernel upgrade brings for the device – especially regarding switchdev. Is the switching ASIC better supported now?

File: packages/platforms/accton/arm64/as4224/src/modules-5_6/arm64-accton-as4224-cpld.c (hunk 2 above)
Comment (brandonchuang) [reply packages/platforms/accton/arm64/as4224/src/modules-5_6/arm64-accton-as4224-cpld.c]: Rollback to previous version since this is not related to kernel 5.10.

File: packages/base/any/kernels/5.10-lts/patches/0012-accton-as4224.patch (hunk 3 above)
Comment (brandonchuang) [reply packages/base/any/kernels/5.10-lts/patches/0012-accton-as4224.patch:84]: Already added this changes in a separate commit. Please help check.

**Comment by brandonchuang** [comment]
> Thank you for submitting the changes. Currently, it’s hard to review though, as it’s only one big commit, and it looks like there are small changes in the patches, which I do not fully understand, how they are related to updating the Linux kernel version.
> 
> Also, a description would be nice, what new features the Linux kernel upgrade brings for the device – especially regarding switchdev. Is the switching ASIC better supported now?

You are right. The switchdev should support kernel 5.10 before merge the changes.

File: packages/platforms/accton/arm64/as4224/src/modules-5_10/arm64-accton-as4224-psu.c (hunk 1 above)
Comment (brandonchuang) [reply packages/platforms/accton/arm64/as4224/src/modules-5_10/arm64-accton-as4224-psu.c]: Modified to follow Linux kernel coding style.

**Review by sonoble — APPROVED** [review]

**Review by Mickey201 — DISMISSED** [review]
this patch was also confirmed with Marvell Switchdev driver.

**Comment by paulmenzel** [comment]
@brandonchuang, thank you for the updates. Fixup commits really make it hard to understand, what is happening. Can you please squash them where necessary? From [*How to squash git commits*][1]:

> Squashing is mainly used to condense a large number of commits to make it to a small number of meaningful commits. So that we can make the git history clearer.

Also, follow [*How to Write a Git Commit Message*][2] and amend *[as4224/as5114] Porting latest DTS patch from kernel 5.6 to 5.10* to use imperative mood:

> [as4224/as5114] Port latest DTS patch from kernel 5.6 to 5.10

[1]: https://gitbetter.substack.com/p/how-to-squash-git-commits
[2]: https://chris.beams.io/posts/git-commit/

**Comment by paulmenzel** [comment]
> this patch was also confirmed with Marvell Switchdev driver.

@Mickey201, is the test protocol documented somewhere? What switchdev driver was used? 2.0 or 2.8? Was it tested on the AS4224 and AS5114?

**Comment by Mickey201** [comment]
Hi Paul, The test protocols are all based on Marvell regression reports. Currently, there’s no official mechanism that can enable review it. It all used as temporary solution till the official DENT testing will be introduced.
Switchdev version 2.8.0 + LK v5.10
Tested over both AS4224/AS5114 platforms.

Regards,
Mickey.

**Comment by brandonchuang** [comment]
I tried to squash commits into one, but after force push, it contains other commits which is not related to this one.
Should I create another PR to fix this?

**Comment by chenglin-tsai** [comment]
> I tried to squash commits into one, but after force push, it contains other commits which is not related to this one.
> Should I create another PR to fix this?

@brandonchuang, If I were you, I would like to update the current branch as same as to main branch first, then apply your own changes and do force push again, which doesn't need to create another PR. The detailed steps are as following:
1. Save current changes as patches with ```git format-patch -n```, n is 1 to N
2. Remove commits not in main branch with ```git reset HEAD^n```, ```git stash```, and ```git stash drop```
3. Update kernel_5.10 branch as same as to main branch with ```git pull <upstream> main [--rebase]```
4. Apply your patches created in step1 with ```git apply <patch-name>``` and commit them with ```git commit -s -S```.
5. Force push to github with ```git push -f```

Maybe there are other better ways to fix it, so the above steps are just for your reference.

**Comment by chenglin-tsai** [comment]
To insert Marvell prestera drivers in kernel 5.10, it should be using ```self.modprobe(`prestera_pci`)``` \[1\]\[2\].
https://github.com/dentproject/dentOS/blob/557ad689c88ca84cdf664500c9319ae78540c7d4/packages/platforms/accton/arm64/as4224/as5114-48x/platform-config/r0/src/python/arm64_accton_as5114_48x_r0/__init__.py#L52

\[1\]: https://github.com/dentproject/dentOS/pull/43#issuecomment-777704551
\[2\]: https://github.com/dentproject/dentOS/pull/51

**Comment by brandonchuang** [comment]
> > I tried to squash commits into one, but after force push, it contains other commits which is not related to this one.
> > Should I create another PR to fix this?
> 
> @brandonchuang, If I were you, I would like to update the current branch as same as to main branch first, then apply your own changes and do force push again, which doesn't need to create another PR. The detailed steps are as following:
> 
> 1. Save current changes as patches with `git format-patch -n`, n is 1 to N
> 2. Remove commits not in main branch with `git reset HEAD^n`, `git stash`, and `git stash drop`
> 3. Update kernel_5.10 branch as same as to main branch with `git pull <upstream> main [--rebase]`
> 4. Apply your patches created in step1 with `git apply <patch-name>` and commit them with `git commit -s -S`.
> 5. Force push to github with `git push -f`
> 
> Maybe there are other better ways to fix it, so the above steps are just for your reference.

I tried and it works.
Thank you for your assistance with this matter.

**Comment by brandonchuang** [comment]
> To insert Marvell prestera drivers in kernel 5.10, it should use `` self.modprobe(`prestera_pci`) `` [1][2].
> https://github.com/dentproject/dentOS/blob/557ad689c88ca84cdf664500c9319ae78540c7d4/packages/platforms/accton/arm64/as4224/as5114-48x/platform-config/r0/src/python/arm64_accton_as5114_48x_r0/__init__.py#L52
> 
> [1]: [#43 (comment)](https://github.com/dentproject/dentOS/pull/43#issuecomment-777704551)
> [2]: #51

Done

**Comment by chenglin-tsai** [comment]
> > > I tried to squash commits into one, but after force push, it contains other commits which is not related to this one.
> > > Should I create another PR to fix this?
> > 
> > 
> > @brandonchuang, If I were you, I would like to update the current branch as same as to main branch first, then apply your own changes and do force push again, which doesn't need to create another PR. The detailed steps are as following:
> > 
> > 1. Save current changes as patches with `git format-patch -n`, n is 1 to N
> > 2. Remove commits not in main branch with `git reset HEAD^n`, `git stash`, and `git stash drop`
> > 3. Update kernel_5.10 branch as same as to main branch with `git pull <upstream> main [--rebase]`
> > 4. Apply your patches created in step1 with `git apply <patch-name>` and commit them with `git commit -s -S`.
> > 5. Force push to github with `git push -f`
> > 
> > Maybe there are other better ways to fix it, so the above steps are just for your reference.
> 
> I tried and it works.
> Thank you for your assistance with this matter.

@brandonchuang It is good to see these steps helps. By the way, do Marvell firmware 2.8.0 and prestera drivers in kernel 5.10 run OK on your boxes? It would be better if you have the test logs.

**Review by taraschornyiplv — APPROVED** [review]

File: packages/base/any/kernels/5.10-lts/patches/0012-accton-as4224.patch (hunk 3 above)
Comment (paulmenzel) [reply packages/base/any/kernels/5.10-lts/patches/0012-accton-as4224.patch:84]: Sorry, I must have missed it. Can you explain again, why this is changed, that means, what problem is fixed?

Comment (Mickey201) [reply packages/base/any/kernels/5.10-lts/patches/0012-accton-as4224.patch:84]: It fixes wrong CPLD implementation. This is the RX_LOSS signal from SFP towards the CPLD. These changes were confirmed on privet branch with Marvell Switchdev drivers + Accton AS5114 platforms.

**Review by Mickey201 — DISMISSED** [review]

**Comment by Mickey201** [comment]
Hi @brandonchuang ,
Please sign your commit. it can't be merged without it.
For more information for the signing - please check the below link:
https://docs.github.com/en/github/authenticating-to-github/about-commit-signature-verification

**Comment by paulmenzel** [comment]
The commit is signed off.

**Comment by paulmenzel** [comment]
Maybe @sonoble’s merge commit is the problem? @brandonchuang, you might then rebase the commit on the master branch `git rebase origin/master`, and force-push again.

**Comment by Mickey201** [comment]
This 872ca89 commit is not marked 'verified' .
Any chance it is related to: "GPG commit signature verification" ?

**Comment by sonoble** [comment]
It doesn't show verified on Brandon's commit. I can merge but I think that is the issue as my commit shows verified. 

Screenshot_20210304-085158
Screenshot_20210304-085128

**Comment by paulmenzel** [comment]
Indeed, you are spot on. It’s not about the Signed-off-by line, but cryptographically signing the commit.

**Comment by brandonchuang** [comment]
> It doesn't show verified on Brandon's commit. I can merge but I think that is the issue as my commit shows verified.
> 
> Screenshot_20210304-085158
> Screenshot_20210304-085128

I tried again but it still doesn't show verified.
Any idea?

**Comment by Mickey201** [comment]
Hi @brandonchuang ,
Did you followed the below link:
https://docs.github.com/en/github/authenticating-to-github/about-commit-signature-verification#gpg-commit-signature-verification
We think this is what prevents the "verified" marking.

**Review by sonoble — APPROVED** [review]

**Review by Mickey201 — APPROVED** [review]
